### ISSUE TITLE
docker:dind can't initialize iptables table `nat': Table does not exist

### ISSUE THREAD

**Issue by bobesz** [docker-library/docker#350]
```console
Generating RSA private key, 4096 bit long modulus (2 primes)
..........................++++
.......................++++
e is 65537 (0x010001)
Generating RSA private key, 4096 bit long modulus (2 primes)
.......................++++
..........................++++
e is 65537 (0x010001)

Signature ok
subject=CN = docker:dind server
Getting CA Private Key
/certs/server/cert.pem: OK

Signature ok
subject=CN = docker:dind client
Getting CA Private Key
/certs/client/cert.pem: OK
time="2022-01-18T18:56:44.278906742Z" level=info msg="Starting up"
time="2022-01-18T18:56:44.281106798Z" level=warning msg="could not change group /var/run/docker.sock to docker: group docker not found"
time="2022-01-18T18:56:44.282765707Z" level=info msg="libcontainerd: started new containerd process" pid=62
time="2022-01-18T18:56:44.282866755Z" level=info msg="parsed scheme: \"unix\"" module=grpc
time="2022-01-18T18:56:44.282890610Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
time="2022-01-18T18:56:44.282933489Z" level=info msg="ccResolverWrapper: sending update to cc: {[{unix:///var/run/docker/containerd/containerd.sock  <nil> 0 <nil>}] <nil> <nil>}" module=grpc
time="2022-01-18T18:56:44.283124724Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
time="2022-01-18T18:56:44.301466160Z" level=info msg="starting containerd" revision=7b11cfaabd73bb80907dd23182b9347b4245eb5d version=v1.4.12
time="2022-01-18T18:56:44.331837231Z" level=info msg="loading plugin \"io.containerd.content.v1.content\"..." type=io.containerd.content.v1
time="2022-01-18T18:56:44.332086213Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.aufs\"..." type=io.containerd.snapshotter.v1
time="2022-01-18T18:56:44.343987441Z" level=info msg="skip loading plugin \"io.containerd.snapshotter.v1.aufs\"..." error="aufs is not supported (modprobe aufs failed: exit status 1 \"ip: can't find device 'aufs'\\nmodprobe: can't change directory to '/lib/modules': No such file or directory\\n\"): skip plugin" type=io.containerd.snapshotter.v1
time="2022-01-18T18:56:44.344051530Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.btrfs\"..." type=io.containerd.snapshotter.v1
time="2022-01-18T18:56:44.344452885Z" level=info msg="skip loading plugin \"io.containerd.snapshotter.v1.btrfs\"..." error="path /var/lib/docker/containerd/daemon/io.containerd.snapshotter.v1.btrfs (ext4) must be a btrfs filesystem to be used with the btrfs snapshotter: skip plugin" type=io.containerd.snapshotter.v1
time="2022-01-18T18:56:44.344491948Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.devmapper\"..." type=io.containerd.snapshotter.v1
time="2022-01-18T18:56:44.344577927Z" level=warning msg="failed to load plugin io.containerd.snapshotter.v1.devmapper" error="devmapper not configured"
time="2022-01-18T18:56:44.344622220Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.native\"..." type=io.containerd.snapshotter.v1
time="2022-01-18T18:56:44.344724319Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.overlayfs\"..." type=io.containerd.snapshotter.v1
time="2022-01-18T18:56:44.345014268Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.zfs\"..." type=io.containerd.snapshotter.v1
time="2022-01-18T18:56:44.345229538Z" level=info msg="skip loading plugin \"io.containerd.snapshotter.v1.zfs\"..." error="path /var/lib/docker/containerd/daemon/io.containerd.snapshotter.v1.zfs must be a zfs filesystem to be used with the zfs snapshotter: skip plugin" type=io.containerd.snapshotter.v1
time="2022-01-18T18:56:44.345264603Z" level=info msg="loading plugin \"io.containerd.metadata.v1.bolt\"..." type=io.containerd.metadata.v1
time="2022-01-18T18:56:44.345320166Z" level=warning msg="could not use snapshotter devmapper in metadata plugin" error="devmapper not configured"
time="2022-01-18T18:56:44.345344161Z" level=info msg="metadata content store policy set" policy=shared
time="2022-01-18T18:56:44.348669564Z" level=info msg="loading plugin \"io.containerd.differ.v1.walking\"..." type=io.containerd.differ.v1
time="2022-01-18T18:56:44.348730448Z" level=info msg="loading plugin \"io.containerd.gc.v1.scheduler\"..." type=io.containerd.gc.v1
time="2022-01-18T18:56:44.348823159Z" level=info msg="loading plugin \"io.containerd.service.v1.introspection-service\"..." type=io.containerd.service.v1
time="2022-01-18T18:56:44.348875386Z" level=info msg="loading plugin \"io.containerd.service.v1.containers-service\"..." type=io.containerd.service.v1
time="2022-01-18T18:56:44.348892719Z" level=info msg="loading plugin \"io.containerd.service.v1.content-service\"..." type=io.containerd.service.v1
time="2022-01-18T18:56:44.348908218Z" level=info msg="loading plugin \"io.containerd.service.v1.diff-service\"..." type=io.containerd.service.v1
time="2022-01-18T18:56:44.348925510Z" level=info msg="loading plugin \"io.containerd.service.v1.images-service\"..." type=io.containerd.service.v1
time="2022-01-18T18:56:44.348962359Z" level=info msg="loading plugin \"io.containerd.service.v1.leases-service\"..." type=io.containerd.service.v1
time="2022-01-18T18:56:44.348980202Z" level=info msg="loading plugin \"io.containerd.service.v1.namespaces-service\"..." type=io.containerd.service.v1
time="2022-01-18T18:56:44.349003765Z" level=info msg="loading plugin \"io.containerd.service.v1.snapshots-service\"..." type=io.containerd.service.v1
time="2022-01-18T18:56:44.349033149Z" level=info msg="loading plugin \"io.containerd.runtime.v1.linux\"..." type=io.containerd.runtime.v1
time="2022-01-18T18:56:44.349268787Z" level=info msg="loading plugin \"io.containerd.runtime.v2.task\"..." type=io.containerd.runtime.v2
time="2022-01-18T18:56:44.349418175Z" level=info msg="loading plugin \"io.containerd.monitor.v1.cgroups\"..." type=io.containerd.monitor.v1
time="2022-01-18T18:56:44.350147058Z" level=info msg="loading plugin \"io.containerd.service.v1.tasks-service\"..." type=io.containerd.service.v1
time="2022-01-18T18:56:44.350201479Z" level=info msg="loading plugin \"io.containerd.internal.v1.restart\"..." type=io.containerd.internal.v1
time="2022-01-18T18:56:44.350310451Z" level=info msg="loading plugin \"io.containerd.grpc.v1.containers\"..." type=io.containerd.grpc.v1
time="2022-01-18T18:56:44.350328285Z" level=info msg="loading plugin \"io.containerd.grpc.v1.content\"..." type=io.containerd.grpc.v1
time="2022-01-18T18:56:44.350343263Z" level=info msg="loading plugin \"io.containerd.grpc.v1.diff\"..." type=io.containerd.grpc.v1
time="2022-01-18T18:56:44.350357278Z" level=info msg="loading plugin \"io.containerd.grpc.v1.events\"..." type=io.containerd.grpc.v1
time="2022-01-18T18:56:44.350372036Z" level=info msg="loading plugin \"io.containerd.grpc.v1.healthcheck\"..." type=io.containerd.grpc.v1
time="2022-01-18T18:56:44.350388727Z" level=info msg="loading plugin \"io.containerd.grpc.v1.images\"..." type=io.containerd.grpc.v1
time="2022-01-18T18:56:44.350403083Z" level=info msg="loading plugin \"io.containerd.grpc.v1.leases\"..." type=io.containerd.grpc.v1
time="2022-01-18T18:56:44.350417290Z" level=info msg="loading plugin \"io.containerd.grpc.v1.namespaces\"..." type=io.containerd.grpc.v1
time="2022-01-18T18:56:44.350437267Z" level=info msg="loading plugin \"io.containerd.internal.v1.opt\"..." type=io.containerd.internal.v1
time="2022-01-18T18:56:44.350757241Z" level=info msg="loading plugin \"io.containerd.grpc.v1.snapshots\"..." type=io.containerd.grpc.v1
time="2022-01-18T18:56:44.350785173Z" level=info msg="loading plugin \"io.containerd.grpc.v1.tasks\"..." type=io.containerd.grpc.v1
time="2022-01-18T18:56:44.350802886Z" level=info msg="loading plugin \"io.containerd.grpc.v1.version\"..." type=io.containerd.grpc.v1
time="2022-01-18T18:56:44.350817323Z" level=info msg="loading plugin \"io.containerd.grpc.v1.introspection\"..." type=io.containerd.grpc.v1
time="2022-01-18T18:56:44.351136676Z" level=info msg=serving... address=/var/run/docker/containerd/containerd-debug.sock
time="2022-01-18T18:56:44.351231221Z" level=info msg=serving... address=/var/run/docker/containerd/containerd.sock.ttrpc
time="2022-01-18T18:56:44.351310047Z" level=info msg=serving... address=/var/run/docker/containerd/containerd.sock
time="2022-01-18T18:56:44.351335094Z" level=info msg="containerd successfully booted in 0.050865s"
time="2022-01-18T18:56:44.360172343Z" level=info msg="parsed scheme: \"unix\"" module=grpc
time="2022-01-18T18:56:44.360205294Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
time="2022-01-18T18:56:44.360232785Z" level=info msg="ccResolverWrapper: sending update to cc: {[{unix:///var/run/docker/containerd/containerd.sock  <nil> 0 <nil>}] <nil> <nil>}" module=grpc
time="2022-01-18T18:56:44.360242333Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
time="2022-01-18T18:56:44.361450035Z" level=info msg="parsed scheme: \"unix\"" module=grpc
time="2022-01-18T18:56:44.361481103Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
time="2022-01-18T18:56:44.361509154Z" level=info msg="ccResolverWrapper: sending update to cc: {[{unix:///var/run/docker/containerd/containerd.sock  <nil> 0 <nil>}] <nil> <nil>}" module=grpc
time="2022-01-18T18:56:44.361522459Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
time="2022-01-18T18:56:44.395337335Z" level=warning msg="Your kernel does not support cgroup blkio weight"
time="2022-01-18T18:56:44.395387328Z" level=warning msg="Your kernel does not support cgroup blkio weight_device"
time="2022-01-18T18:56:44.395686954Z" level=info msg="Loading containers: start."
time="2022-01-18T18:56:44.418135857Z" level=warning msg="Running modprobe bridge br_netfilter failed with message: ip: can't find device 'bridge'\nbridge                200704  1 br_netfilter\nstp                    16384  1 bridge\nllc                    16384  2 bridge,stp\nip: can't find device 'br_netfilter'\nbr_netfilter           24576  0 \nbridge                200704  1 br_netfilter\nmodprobe: can't change directory to '/lib/modules': No such file or directory\n, error: exit status 1"
time="2022-01-18T18:56:44.419938724Z" level=warning msg="Running iptables --wait -t nat -L -n failed with message: `modprobe: can't change directory to '/lib/modules': No such file or directory\niptables v1.8.7 (legacy): can't initialize iptables table `nat': Table does not exist (do you need to insmod?)\nPerhaps iptables or your kernel needs to be upgraded.`, error: exit status 3"
time="2022-01-18T18:56:44.450493337Z" level=info msg="stopping event stream following graceful shutdown" error="context canceled" module=libcontainerd namespace=moby
time="2022-01-18T18:56:44.450625322Z" level=info msg="stopping healthcheck following graceful shutdown" module=libcontainerd
time="2022-01-18T18:56:44.450693268Z" level=info msg="stopping event stream following graceful shutdown" error="context canceled" module=libcontainerd namespace=plugins.moby
time="2022-01-18T18:56:45.451864941Z" level=warning msg="grpc: addrConn.createTransport failed to connect to {unix:///var/run/docker/containerd/containerd.sock  <nil> 0 <nil>}. Err :connection error: desc = \"transport: Error while dialing dial unix:///var/run/docker/containerd/containerd.sock: timeout\". Reconnecting..." module=grpc
failed to start daemon: Error initializing network controller: error obtaining controller instance: failed to create NAT chain DOCKER: iptables failed: iptables -t nat -N DOCKER: modprobe: can't change directory to '/lib/modules': No such file or directory
iptables v1.8.7 (legacy): can't initialize iptables table `nat': Table does not exist (do you need to insmod?)
Perhaps iptables or your kernel needs to be upgraded.
 (exit status 3)
```

**Comment by tianon** [comment]
Can you provide a bit more detail?  What's your host environment like (distro, Docker version, kernel version, etc)?  What flags are you adding to your `docker run` line when you run the container?

**Comment by ricardo-kagawa** [comment]
Hi, I am also getting this error, so maybe my specific details may help.

I originally got this error using the `stable-dind` image (I guess [it should be](https://hub.docker.com/layers/docker/library/docker/stable-dind/images/sha256-2a7e15ecad4d18a6bb71b1b828dc7bb1b1a6c890e7420a89726a0649e94e8fb1?context=explore) 19.03.14) in a Kubernetes cluster version 1.21.5 with Oracle Linux 8.4 nodes. It was working fine previously also with `stable-dind` image in cluster version 1.21.5, but with Oracle Linux 7.9.

In both cases, container is running in privileged mode (as instructed), but with default network mode (I'm not sure which that would be for Kubernetes, but definitely not `host` mode).

Looking at what might have changed in the distro, it seems they have deprecated `iptables` in favor of `firewalld`, which might be the cause for this error. But from [Docker documentation](https://docs.docker.com/network/iptables/), version 20 should be able to use `firewalld` automatically, so I have tried the `20-dind` image as well, without success.

I can provide more details on the current setup if you need, but the old setup is already gone. The logs are very similar to the one already posted, so I'm not posting mine unless asked.

**Comment by tianon** [comment]
(If my hunch is right, then `modprobe ip_tables` will probably fix it, but I'm not sure how that will interact with the host -- probably not at all given it'll only be used inside the container network namespaces.)

**Comment by ricardo-kagawa** [comment]
Haha sorry, I thought it would take a while for you to answer, and came to the conclusion I should rollback and get going with my day. Maybe I'll try to boot a separate host to test it later, instead of messing with my preproduction servers.

But now that you have mentioned it, my logs were actually slightly different there. It mentioned it could not find `/var/libraries` (or something like that) to load `iptables` module. Just missed that among all the other messages. My logs were also wiped along with the cluster node (maybe I should have saved a copy, just in case... too late now).

Oracle's docs do mention it uses `nftables` instead of `iptables` now.

**Comment by ricardo-kagawa** [comment]
Hello again, just figured a way to reproduce the failing container without having to recreate the cluster node.

In a freshly booted host, `lsmod | grep ip_tables` turned out empty. `lsmod` itself is not empty.

`modprobe ip_tables` from inside the container seems to fail with code 1:
```
ip: can't find device 'ip_tables'
modprobe: can't change directory to '/lib/modules': No such file or directory
```

But actually, `lsmod | grep ip_tables` is then not empty anymore.

In fact, starting another container in the same node now seems to work now, which might answer your question about "how it will interact with the host". Although I did use `kubectl debug node/$NODE` here to try to use a privileged container, and I'm not quite sure how privileged it could be (the docs say it would have access to the host namespace and filesystem, but I have not mounted anything and it seemed like I was in the container's file system instead of the host's; I'm not sure about the namespace either).

And since I'm already here, I could also add the output of `uname -a`:
```
Linux test-docker 5.4.17-2102.206.1.el8uek.x86_64 #2 SMP Wed Oct 6 16:24:16 PDT 2021 x86_64 Linux
```

I could try again later if this info looks too erratic, but then I probably should have a better testing plan to get more concrete results. Maybe use `kubectl run --privileged` instead, which should be closer to how an actual container would run, and only then see the effects on the host? (Of course, with a fresh host.)

**Comment by tianon** [comment]
Ah, nice!  That gives a good clue as to how we can fix this.  In the interest of detecting it properly, would you be willing to do `rmmod ip_tables` (probably on the host), run the `docker:dind` container with a command like `sh` and give me the output of something like `iptables -nvL`?  Maybe even `iptables -t nat -nvL`?  (Trying to find the best way to detect this case so we can apply the workaround appropriately to avoid extra error messages for unaffected users which would be harmless but also meaningless and unnecessarily alarming.)

**Comment by ricardo-kagawa** [comment]
I'm running this in a fresh cluster node, using `kubectl run --privileged` this time, to be on a safer side.

`iptables -nvL` exits with code 3 with output:

```
modprobe: can't change directory to '/lib/modules': No such file or directory
iptables v1.8.7 (legacy): can't initialize iptables table `filter': Table does not exist (do you need to insmod?)
Perhaps iptables or your kernel needs to be upgraded.
```

`iptables -t nat -nvL` yields the same exit code and output.

(This output is identical to the last logged lines, by the way.)

---

Since now I'm running things with `kubectl run --privileged`, it might be worth testing `modprobe ip_tables` again to see how it affects the host.

So again `modprobe ip_tables` exits with code 1 with output:

```
ip: can't find device 'ip_tables'
ip_tables              28672  0 
modprobe: can't change directory to '/lib/modules': No such file or directory
```

But now `lsmod | grep ip_tables` finds the module:

```
ip_tables              28672  0 
```

And a new container (not the same I ran `modprobe`) now works in the same node. I have checked that it didn't work before.

This very much confirms that the host will be affected by `modprobe` in this particular system configuration. It might be worth testing with other systems before making any strong conclusions though.
